### ISSUE TITLE
Tensor transfer between gpus doesnt work

### ISSUE THREAD

**Issue by pnsuau** [pytorch/pytorch#70048]
### 🐛 Describe the bug

When transferring a tensor from a gpu to another, weights aren't right on the second gpu whereas it works from cpu to gpu and from gpu to cpu.

```
python3
a = torch.Tensor([0.6,0.6])
b = a.to("cuda:0")
print(b.to("cuda:1"))
```
Output:
```
tensor([0., 0.], device='cuda:1')
```

Expected output
```
tensor([0.6, 0.6], device='cuda:1')
```

### Versions

Collecting environment information...
PyTorch version: 1.10.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 26 2021, 20:14:08)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-89-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: NVIDIA RTX A5000
GPU 1: NVIDIA RTX A5000
GPU 2: NVIDIA RTX A5000
GPU 3: NVIDIA RTX A5000

Nvidia driver version: 495.44
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.17.4
[pip3] torch==1.10.1+cu113
[pip3] torchaudio==0.10.1+cu113
[pip3] torchfile==0.1.0
[pip3] torchvision==0.11.2+cu113
[pip3] torchviz==0.0.2
[conda] Could not collect

**Comment by mruberry** [comment]
This does not reproduce for me on a recent build of PyTorch -- it may be a hardware specific issue.

**Comment by Zefan-Cai** [comment]
Hi, did you find the sollution of this issue?

**Comment by beniz** [comment]
@Zefan-Cai Absolutely, we eventually solved it. This was an IOMMU hardware issue, to solve it, you want to set the `ommu=soft` flag to the linux kernel boot configuration line in grub.
See https://unix.stackexchange.com/questions/469153/what-are-the-implication-of-iommu-soft for an overview (unrelated to GPUs). This fixed the issue on all our machines with the same motherboard.

**Comment by Assassin321** [comment]
Thank you so much for your message! I tried setting the iommu=soft in /etc/default/grub by modifying GRUB_CMDLINE_LINUX_DEFAULT to "GRUB_CMDLINE_LINUX_DEFAULT=quiet splash iommu=soft". However, even after rebooting the server, it still doesn't seem to work. The same tensor transferred from cuda:0 to cuda:1 is resulting in different values. Is there anything else that needs to be done to resolve this hardware issue?

**Comment by Zefan-Cai** [comment]
> Thank you so much for your message! I tried setting the iommu=soft in /etc/default/grub by modifying GRUB_CMDLINE_LINUX_DEFAULT to "GRUB_CMDLINE_LINUX_DEFAULT=quiet splash iommu=soft". However, even after rebooting the server, it still doesn't seem to work. The same tensor transferred from cuda:0 to cuda:1 is resulting in different values. Is there anything else that needs to be done to resolve this hardware issue?

I have a same situation.

**Comment by beniz** [comment]
You need to run `update-grub` afterwards, if you haven't done it already, then reboot. If this doesn't work then your issue is different than ours :/

**Comment by muellerzr** [comment]
This solution does not work for us on the latest pytorch version (I checked grub and this value was already set).

Env is 2x4090 GPUs. Simple code reproduction:

```python
import torch 
tensor_cpu = torch.tensor([[1,2,3]])
print(tensor_cpu)
tensor_1 = tensor_cpu.to(1)
print(tensor_1)
tensor_1_to_0 = tensor_1.to(0)
print(tensor_1_to_0)
```
